c10/xpu/XPUStream.cpp

```diff
@@ -0,0 +1,266 @@
+#include <c10/util/CallOnce.h>
+#include <c10/util/irange.h>
+#include <c10/xpu/XPUException.h>
+#include <c10/xpu/XPUStream.h>
+
+#include <atomic>
+#include <deque>
+#include <mutex>
+#include <vector>
+
+namespace c10::xpu {
+namespace {
+
+// Global stream state and constants
+c10::once_flag init_flag;
+DeviceIndex num_gpus = -1;
+constexpr int kStreamsPerPoolBits = 5;
+constexpr int kStreamsPerPool = 1 << kStreamsPerPoolBits;
+constexpr int kStreamTypeBits = 3;
+
+// The SYCL queue pools are lazily initialized when the first queue is requested
+// for a device. The device flags track the initialization of each device. When
+// a queue is requested, the next queue in the pool to be returned in a
+// round-robin fashion, see Note [Stream Management].
+std::deque<c10::once_flag> device_flags;
+std::vector<std::array<
+    std::array<std::unique_ptr<sycl::queue>, kStreamsPerPool>,
+    max_compile_time_stream_priorities>>
+    streams;
+std::deque<
+    std::array<std::atomic<uint32_t>, max_compile_time_stream_priorities>>
+    priority_counters;
+
+thread_local std::unique_ptr<StreamId[]> current_streams = nullptr;
+
+// Note [StreamId assignment]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~
+// How do we assign stream IDs?
+//
+// -- 57 bits --  -- 5 bits -----  -- 3 bits --
+//     zeros      StreamIdIndex    StreamIdType
+//
+// Where StreamIdType:
+//  000 = normal priority queue
+//  001 = high priority queue
```

Comment:
SYCL has no high prio queue. Should we remove this queue type?

Reply:
BTW, in IPEX, we have below.

// Where QueueType:
//  000 = UNUSED
//  001 = reserved queue

Reply:
Looks like this dese support as SYCL extension.
But we never used it before... Should we confirm with SYCL team?

https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_queue_priority.asciidoc

Reply:
I have checked `normal priority` is the default priority. see https://github.com/intel/llvm/pull/7523/files#diff-25fb98dfc3242dac5e5f1db93388f508fe3a5a617a3eadebbaa587d886112197R319

Reply:
we should follow pytorch style. pls check in pytorch if has same coding?

Reply:
If we do not use it now, we'd better remove it and add it back on demand.

Reply:
We may only keep the normal priority queue and leave the priority type bits for future extension.